src/rust/src/orchard_ffi/incremental_sinsemilla_tree_ffi.rs

```diff
@@ -0,0 +1,285 @@
+    stream: Option<StreamObj>,
+    read_cb: Option<ReadCb>,
+) -> *mut BridgeTree<OrchardIncrementalTreeDigest> {
+    let reader = CppStreamReader::from_raw_parts(stream, read_cb.unwrap());
+
+    match bincode::deserialize_from(reader) {
+        Ok(parsed) => Box::into_raw(Box::new(parsed)),
+        Err(e) => {
+            error!("Failed to parse Orchard bundle: {}", e);
+            ptr::null_mut()
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_serialize(
+    tree: *const BridgeTree<OrchardIncrementalTreeDigest>,
+    stream: Option<StreamObj>,
+    write_cb: Option<WriteCb>,
+) -> bool {
+    let tree = unsafe {
+        tree.as_ref()
+            .expect("Orchard tree pointer may not be null.")
+    };
+
+    let writer = CppStreamWriter::from_raw_parts(stream, write_cb.unwrap());
+    match bincode::serialize_into(writer, tree) {
+        Ok(()) => true,
+        Err(e) => {
+            error!("{}", e);
+            false
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_append_bundle(
+    tree: *mut BridgeTree<OrchardIncrementalTreeDigest>,
+    bundle: *const orchard::Bundle<Authorized, Amount>,
+) -> bool {
+    let tree = unsafe {
+        tree.as_mut()
+            .expect("Orchard note commitment tree pointer may not be null.")
+    };
+    if let Some(bundle) = unsafe { bundle.as_ref() } {
+        for action in bundle.actions().iter() {
+            if !tree.append(&OrchardIncrementalTreeDigest::from_cmx(action.cmx())) {
+                error!("Orchard note commitment tree is full.");
+                return false;
+            }
+        }
+
+        true
+    } else {
+        false
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_checkpoint(
+    tree: *mut BridgeTree<OrchardIncrementalTreeDigest>,
+) {
+    let tree = unsafe {
+        tree.as_mut()
+            .expect("Orchard tree pointer may not be null.")
+    };
+
+    tree.checkpoint()
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_rewind(
+    tree: *mut BridgeTree<OrchardIncrementalTreeDigest>,
+) -> bool {
+    let tree = unsafe {
+        tree.as_mut()
+            .expect("Orchard tree pointer may not be null.")
+    };
+
+    tree.rewind()
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_root(
+    tree: *mut BridgeTree<OrchardIncrementalTreeDigest>,
+    root_ret: *mut [u8; 32],
+) -> bool {
+    let root_ret = unsafe {
+        root_ret
+            .as_mut()
+            .expect("Cannot return to the null pointer.")
+    };
+    if let Some(tree) = unsafe { tree.as_ref() } {
+        if let Some(root) = tree.root().to_bytes() {
+            root_ret.copy_from_slice(&root);
+            true
+        } else {
+            false
+        }
+    } else {
+        false
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn incremental_sinsemilla_tree_empty_root(depth: usize, root_ret: *mut [u8; 32]) {
```

Comment:
Where is this API used?

Reply:
Used here: https://github.com/zcash/zcash/pull/5233/files#diff-31ef5cdda38b86238aa5fa0745b7843ebd43f11d65583fb9cfec23f7225e5f77R339

Reply:
Will this API be used with any depth other than the Orchard depth?

Reply:
Nope, I can remove it.